app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java

```diff
@@ -113,113 +81,67 @@ public void onScanButtonClick(final View view) {
-                    .setView(editName)
-                    .setPositiveButton(getString(R.string.done), enClickListener);
-            lastDialog = builder.create();
-            lastDialog.show();
-        } catch (IllegalArgumentException e) {
-            LOGGER.error("Illegal argument exception in addNewContact", e);
-        }
-    }
-
-
-
-
     /**
-     * Shows the QR Code in an alert dialog popup screen
-     *
-     * @param qrCode The QR code to be shown.
+     * Shows the QR Code in the activity.
      */
     @SuppressLint("InflateParams")
-    private void showQRCode(final Bitmap qrCode) {
-
-        final LayoutInflater li = LayoutInflater.from(this);
-        final View myView = li.inflate(R.layout.qrcode, null);
+    private void drawQRCode() {
+        final IKey publicKey = profile.getPublicKey();
+        final Bitmap qrCode = QRGenerator.encode(profile.getName() + SEMICOLON + publicKey.getType()
+                + SPACE + publicKey.getName() + SPACE + publicKey.getKey());
 
-        final ImageView imageView = (ImageView) myView.findViewById(R.id.QR_code_image);
+        final ImageView imageView = (ImageView) this.findViewById(R.id.QR_code_image);
         imageView.setImageBitmap(qrCode);
-        final AlertDialog.Builder builder = new AlertDialog.Builder(this)
-                .setView(imageView)
-                .setPositiveButton(getString(R.string.done), new QRCloser());
-
-        ((ViewGroup) imageView.getParent()).removeView(imageView);
-        lastDialog = builder.create();
-        lastDialog.show();
     }
 
-
-    private static final class QRCloser implements DialogInterface.OnClickListener {
-        /**
-         * {@inheritDoc}
-         */
-        @Override
-        public void onClick(final DialogInterface dialog, final int which) {
-            dialog.dismiss();
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onActivityResult(final int requestCode, final int resultCode, final Intent intent) {
+        LOGGER.info("Activity resulted");
+        final IntentResult scanResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
+        if (scanResult == null) {
+            LOGGER.error("No scan result in QR Scanner");
+        } else if (resultCode == RESULT_OK) {
+            final String result = scanResult.getContents();
+            addNewContact(result);
         }
     }
 
-    private final class EditNameClickListener implements DialogInterface.OnClickListener {
-
-        private final EditText editName;
-        private final IKey key;
+    /**
+     * Adds new contact with the scanned key and opens change
+     *
+     * @param qrMessage The information we got from the QR code.
+     */
+    private void addNewContact(final String qrMessage) {
+        try {
+            LOGGER.info("Adding new contact to database");
+            //Name is the first part
+            final String name = qrMessage.split(SEMICOLON)[0];
```

Comment:
Maybe we can use a newline character instead? It is also easier to enforce users not using it through the type of input we use in the `CreateProfileActivity`